Migrations (#136)
kwilteam/docs: docs/node/migrations/zero-downtime-migrations/state-replication.mdx

---
sidebar_position: 2
sidebar_label: "State Replication"
id: state-replication
title: State Replication
description: State Replication
slug: /node/migrations/zero-downtime-migrations/state-replication
---

Zero-Downtime Migrations require the new network to start with the state of the existing network at the start of the migration and then replicate the state changes of the existing network to the new network during the migration period to stay in sync with the existing network.

## State Replication

> State replication is handled by an inbuilt `migrations` listener extension in the Kwil node. The `migrations` listener extension polls for the state changes on the existing network and replicates the changes onto the new network. The `migrations` listener extension can be enabled by setting the below configuration in the `config.toml` and `genesis.json` files.

Reviewer: I think built-in is correct?

```suggestion
State replication is handled by a built-in `migrations` listener extension in the Kwil node. The `migrations` listener extension polls for the state changes on the existing network and replicates the changes onto the new network. The `migrations` listener extension can be enabled by setting the migration-related settings the `config.toml` and `genesis.json` files.
```

`migrate_from` config in the `config.toml` file specifies the JSON-RPC address of the node from the existing network that the new network should replicate the state changes from. 
```toml
[app]

> "migrate_from" = "http://kwild-jsonrpc-server-address:8484"

Reviewer 2: No quotes around `migrate_from` I believe.

Should also mention that the connection should be secure to ensure the received data is not tampered with.

```

The `consensus_params` in the `genesis.json` file specify the start and end block heights that the new network should use to replicate the state changes from the existing network.
```json
{
    "consensus_params": {
        "migration": {
            "start_height": <start_height>,
            "end_height": <end_height>,
        },
    }
}

> ```json

Reviewer 2: Some copy pasta here.

{
    "consensus_params": {
        "migration": {
            "start_height": <start_height>,
            "end_height": <end_height>,
        },
    }

> }

Reviewer: I think this may be a typo.

```
:::note 
Migrations listener extension won't start if the `start_height` and `end_height` are not set in the `genesis.json` file.
:::

## Handling State Divergence

During migration period, as both old and new networks are running concurrently handing user transactions and replicating state changes from the existing network, there is a possibility for the occurance of state divergences between the two networks. 

> The migration process does its best to resolve the conflicts between the two networks using the below conflict resolution principles. However, there might be some schemas where the conflicts cannot be resolved without any loss of data.

Reviewer: "The below" feels a bit awkward. I would just say that there are conflict resolution principles and then list them below.

```suggestion
The migration process does its best to resolve the conflicts between the two networks using a set of conflict resolution principles. However, there might be some schemas where the conflicts cannot be resolved without any loss of data.
```

:::note

> ** Do's and Dont's: ** Operators should migrate to new network as soon as possible and redirect all the user transactions to the new network. Any single user transactions should not be scattered across both new and old chains to avoid any state divergence and should be redirected to a single network as much as possible. If such situation arises, migration process tries its best to handle conflict resolutions using below principles, but for schema actions that doesn't fit the below model the result might be the loss of data.

Reviewer: What does "Any single user transactions should not be scattered across both new and old chains" mean?

Reviewer: The third sentence is repetitive and can be removed: "If such situation arises, migration process tries its best to handle conflict resolutions using below principles, but for schema actions that doesn't fit the below model the result might be the loss of data."

:::

### Conflict Resolution For User Datasets

User Dataset changeset conflicts are handled as follows:

- **Inserts**: Inserts from the old network will attempt to be applied to the new network, and any conflicts will be ignored, in favor of whatever already exists on the new network.
- **Updates**: The new network will check the pre-change value of the proposed update from the old network's changeset. If the new network's record for that primary key matches the pre-change value in the changeset, it will apply the change. If the values are not equal, it will discard the change, in favor of what data exists on the new network
- **Deletes**: Similar to update, if the record matches the pre-change value in the changeset, it will be deleted. Otherwise, no change will be applied.

### Conflict Resolution For Account Spends

All spends from the old network should be replicated to the new network and the conflicts are handled as follows:

> - If a spend from the old network causes a users balance to go below 0, spend entire balance and set it to 0, else spend the amount as per the spend from the old network.

Reviewer: I am unsure what this is trying to say. Could you do a rewrite?
